apps/emqx_machine/src/emqx_cluster_rpc.erl

```diff
@@ -0,0 +1,325 @@
+        {disc_copies, [node()]},
+        {local_content, true},
+        {record_name, cluster_rpc_mfa},
+        {attributes, record_info(fields, cluster_rpc_mfa)}]),
+    ok = ekka_mnesia:create_table(?CLUSTER_COMMIT, [
+        {type, set},
+        {disc_copies, [node()]},
+        {local_content, true},
+        {record_name, cluster_rpc_commit},
+        {attributes, record_info(fields, cluster_rpc_commit)}]);
+mnesia(copy) ->
+    ok = ekka_mnesia:copy_table(cluster_rpc_mfa, disc_copies),
+    ok = ekka_mnesia:copy_table(cluster_rpc_commit, disc_copies).
+
+start_link() ->
+    RetryMs = application:get_env(emqx_machine, cluster_call_retry_interval, 1000),
+    start_link(node(), ?MODULE, RetryMs).
+start_link(Node, Name, RetryMs) ->
+    gen_statem:start_link({local, Name}, ?MODULE, [Node, RetryMs], []).
+
+-spec multicall(Module, Function, Args) -> {ok, TnxId} | {error, Reason} when
+    Module :: module(),
+    Function :: atom(),
+    Args :: [term()],
+    TnxId :: pos_integer(),
+    Reason :: term().
+multicall(M, F, A) ->
+    multicall(M, F, A, timer:minutes(2)).
+
+-spec multicall(Module, Function, Args, Timeout) -> {ok, TnxId} |{error, Reason} when
+    Module :: module(),
+    Function :: atom(),
+    Args :: [term()],
+    TnxId :: pos_integer(),
+    Timeout :: timeout(),
+    Reason :: term().
+multicall(M, F, A, Timeout) ->
+    MFA = {initiate, {M, F, A}},
+    case ekka_rlog:role() of
+        core ->  gen_statem:call(?MODULE, MFA, Timeout);
+        replicant ->
+            case ekka_rlog_status:upstream_node(?COMMON_SHARD) of
+                {ok, Node} -> gen_statem:call({?MODULE, Node}, MFA, Timeout);
+                disconnected -> {error, disconnected}
+            end
+    end.
+
+-spec query(pos_integer()) -> {'atomic', map()} | {'aborted', Reason :: term()}.
+query(TnxId) ->
+    transaction(fun trans_query/1, [TnxId]).
+
+-spec reset() -> reset.
+reset() -> gen_statem:call(?MODULE, reset).
+
+-spec status() -> {'atomic', [map()]} | {'aborted', Reason :: term()}.
+status() ->
+    transaction(fun trans_status/0, []).
+
+%%%===================================================================
+%%% gen_statem callbacks
+%%%===================================================================
+
+%% @private
+init([Node, RetryMs]) ->
+    {ok, _} = mnesia:subscribe({table, ?CLUSTER_MFA, simple}),
+    {ok, ?CATCH_UP, #{node => Node, retry_interval => RetryMs}, ?CATCH_UP_AFTER(0)}.
+
+callback_mode() ->
+    handle_event_function.
+
+%% @private
+format_status(Opt, [_PDict, StateName, Data]) ->
+    #{state => StateName, data => Data , opt => Opt}.
+
+%% @private
+handle_event(state_timeout, catch_up_delay, _State, Data) ->
+    catch_up(Data);
+
+handle_event(info, {mnesia_table_event, {write, #cluster_rpc_mfa{} = MFARec, _AId}}, ?REALTIME, Data) ->
+    handle_mfa_write_event(MFARec, Data);
+handle_event(info, {mnesia_table_event, {write, #cluster_rpc_mfa{}, _ActivityId}}, ?CATCH_UP, _Data) ->
```

Comment:
A lot of events might be buffered during catchup. Perhaps a simpler solution would be to always read data from the table in the same way, and use table events only to trigger transactions?

Reply:
I think 2 ways are feasible, 
now depends on the order of mnesia event, if the current transaction on 3, the order from 4, 5, 6, to 6, 4, 5, will cause this process to restart (but after restart it will still work) @zmstone  what do you think?

Reply:
This is an excellent idea.

Reply:
done

Reply:
> A lot of events might be buffered during catchup. Perhaps a simpler solution would be to always read data from the table in the same way, and use table events only to trigger transactions?
> 
> This way we shouldn't worry about replaying too many events, or having gaps in the replay process
> 
> To describe the idea in more detail: there is no need to make distinction between realtime and catchup states. This process could be a simple "read the next record" -> "execute action" -> "commit" loop, with iteration triggered by mnesia events. The content of the events could be ignored.

I have already changed gen_statem to gen_server to make it statusless, please help review if there are other things not taken into account, thanks :0
@k32